telegram/ext/dispatcher.py

```diff
@@ -154,46 +160,21 @@ class Dispatcher(Generic[CCT, UD, CD, BD]):
     __singleton = None
     logger = logging.getLogger(__name__)
 
-    @overload
-    def __init__(
-        self: 'Dispatcher[CallbackContext[Dict, Dict, Dict], Dict, Dict, Dict]',
-        bot: 'Bot',
-        update_queue: Queue,
-        workers: int = 4,
-        exception_event: Event = None,
-        job_queue: 'JobQueue' = None,
-        persistence: BasePersistence = None,
-    ):
-        ...
-
-    @overload
-    def __init__(
-        self: 'Dispatcher[CCT, UD, CD, BD]',
-        bot: 'Bot',
-        update_queue: Queue,
-        workers: int = 4,
-        exception_event: Event = None,
-        job_queue: 'JobQueue' = None,
-        persistence: BasePersistence = None,
-        context_types: ContextTypes[CCT, UD, CD, BD] = None,
-    ):
-        ...
-
-    def __init__(
-        self,
-        bot: 'Bot',
-        update_queue: Queue,
-        workers: int = 4,
-        exception_event: Event = None,
-        job_queue: 'JobQueue' = None,
-        persistence: BasePersistence = None,
-        context_types: ContextTypes[CCT, UD, CD, BD] = None,
-    ):
-        self.bot = bot
-        self.update_queue = update_queue
-        self.job_queue = job_queue
-        self.workers = workers
-        self.context_types = cast(ContextTypes[CCT, UD, CD, BD], context_types or ContextTypes())
+    def __init__(self, **kwargs: object):
+        if not kwargs.pop('builder_flag', False):
+            warnings.warn(
+                '`Dispatcher` instances should be built via the `DispatcherBuilder`.',
+                UserWarning,
+                stacklevel=2,
+            )
```

Comment:
1. I really prefer that the function signature would not include kwargs, but rather the keyword args we expect.
2. If we want to force keyword args (as we've discussed in the developers group) we can do:
```python
def __init__(self, **, bot, update_queue, ...):
```
3. If we want to issue a warning, we can use `inspect` to check the program stack, and make sure we were called via `UpdaterBuilder`.

Reply:
My intention was to hide the signature as much as possible from the user. With a proper signature, IDEs will still suggest the required parameters for initializing `Updater/Dispatcher` manually …

And when doing that, I thought that I'd have to include `builder_flag` in the signature if I made the signature explicit. (You suggesting to you `inspect` magic is also new for me ;P)

But I'm not too passionate about it …

Reply:
1. **Kwargs, when you know all the possibile keyword arguments, is evil.
Hard to understand the code, leaves much more space for runtime errors and hard to work with an ide (basically it all goes together)
2. Why are we blocking manual construction of the Updater class? Beacsue we decided that this approach is obsolete. That's fine. It's good to block/warn on incorrect usage.
3. There are two approaches here: adding a "magic" argument which validates the code was called by the Builder class (and it bloats the amount of arguments) or, taking advantage of python's (black magic) reflection capabilities. Since this is something that happens only once at the start of the application (in contrary to constant usage during message processing), I don't mind using some of this black magic. So I'm just a bit in favor of it. Anyway, both approaches are valid in my eyes.
4. But back to article 1 above, I prefer a normal function signature. And as noted in my previous comment, we can still force using keyword arguments (as demonstrated in article 2 of that comment).